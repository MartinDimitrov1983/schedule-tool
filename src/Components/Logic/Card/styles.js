import { GREY, LIGHT_GREY } from '../../../Constants';
const styles = {
  wrapper: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    minHeight: '350px',
    minWidth: '101px',
    cursor: 'pointer',
  },
  header: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    margin: '8px',
  },
  day: {
    fontWeight: 'bold',
  },
  body: (isTemplate) => {
    return {
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
      gap: '10px',
      backgroundColor: `${isTemplate ? GREY : LIGHT_GREY}`,
      minHeight: '80%',
      width: '100%',
      height: '100%',
    };
  },
  emptyBody: {
    backgroundColor: LIGHT_GREY,
    minHeight: '80%',
    width: '100%',
  },
};

export default styles;
