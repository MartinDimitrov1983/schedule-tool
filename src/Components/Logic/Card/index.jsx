import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { useDispatch, useSelector } from 'react-redux';
import { Box, Typography } from '@mui/material';
import { DayTimes, Button } from '../..';
import { removeTime } from '../../../Redux/slices/days';
import { shouldDisableUpload } from '../../../Utils';
import { ADD_TIME, DARK_GREY } from '../../../Constants';
import styles from './styles';

const Card = ({ day, date, dayIndex, isTemplate, autoCompleteData }) => {
  const times = useSelector((state) => state.days.times);
  const isAutoComplete = useSelector((state) => state.days.isAutoComplete);
  const [isHoverd, setIsHovered] = useState(
    times[dayIndex]?.length > 0 || autoCompleteData?.length > 0,
  );
  const [isVisible, setIsVisible] = useState(
    times[dayIndex]?.length > 0 ? false : true,
  );
  const [addTime, setAddTime] = useState(
    isAutoComplete ? autoCompleteData : times[dayIndex],
  );
  const dispatch = useDispatch();

  useEffect(() => {
    setAddTime(isAutoComplete ? autoCompleteData : times[dayIndex]);
    setIsHovered(times[dayIndex]?.length > 0 || autoCompleteData?.length > 0);
  }, [autoCompleteData, dayIndex, isAutoComplete, times]);

  const clickCardHandler = () => {
    setIsHovered(true);
  };
  const mouseLeaveHandler = () => {
    setIsVisible(false);
  };
  const mouseEnterHandler = () => {
    setIsVisible(true);
  };

  const addHandler = () => {
    setAddTime((prevState) => {
      const newState = prevState.concat([{ id: `id-${prevState.length + 1}` }]);
      return newState;
    });
  };

  const removeHandler = (id, index) => {
    setAddTime((prevState) => {
      const newState = prevState.filter((time) => time.id !== id);
      return newState;
    });
    dispatch(removeTime({ dayIndex, timeIndex: index }));
  };

  return (
    <Box
      sx={styles.wrapper}
      onClick={clickCardHandler}
      onMouseLeave={mouseLeaveHandler}
      onMouseEnter={mouseEnterHandler}>
      <Box sx={styles.header}>
        <Typography sx={styles.day}>{day}</Typography>
        <Typography>{date}</Typography>
      </Box>
      {isHoverd ? (
        <Box sx={styles.body(isVisible && isTemplate)}>
          {addTime.map((time, index) => {
            return (
              <DayTimes
                key={`${time}-${index}`}
                isTemplate={isTemplate}
                isCompleted={!shouldDisableUpload(times)}
                dayIndex={dayIndex}
                timeIndex={index}
                removeHandler={() => removeHandler(time.id, index)}
                autocompleteData={autoCompleteData?.[index]}
              />
            );
          })}
          {isVisible && isTemplate && (
            <Button
              text={ADD_TIME}
              disabled={false}
              color={DARK_GREY}
              onClick={addHandler}
            />
          )}
        </Box>
      ) : (
        <Box sx={styles.emptyBody} />
      )}
    </Box>
  );
};

Card.propTypes = {
  day: PropTypes.string,
  date: PropTypes.string,
  dayIndex: PropTypes.number,
  isTemplate: PropTypes.bool,
  autoCompleteData: PropTypes.array,
};

export default Card;
