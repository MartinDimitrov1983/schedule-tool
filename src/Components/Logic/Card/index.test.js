import React from 'react';
import { render } from '@testing-library/react';
import * as Redux from 'react-redux';

import Card from './index';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

test('Snapshot Card', () => {
  jest.spyOn(Redux, 'useSelector').mockReturnValue({
    days: {
      times: [['09:00'], []],
      isAutoComplete: false,
    },
  });
  const { asFragment } = render(
    <Card
      day=""
      date=""
      dayIndex={2}
      isTemplate={false}
      autoCompleteData={[]}
    />,
  );
  expect(asFragment()).toMatchSnapshot();
});
