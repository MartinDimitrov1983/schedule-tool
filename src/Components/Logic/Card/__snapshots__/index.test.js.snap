// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Card 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-14t9pl7"
  >
    <div
      class="MuiBox-root css-moyowv"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
      />
      <p
        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      />
    </div>
    <div
      class="MuiBox-root css-3n8u8f"
    />
  </div>
</DocumentFragment>
`;
